Add `pr list --search` and `--author` flags

mislav opened the pull request:
This adds search functionality to `pr list` to complement `issue list --search`.

Since the search backend is now easily available through code, this also adds support for `--author` filter for compatibility with `issue list` #2628 /cc @GeorgeMac 

Followup to https://github.com/cli/cli/pull/3196 /cc @g14a 

Fixes https://github.com/cli/cli/issues/452, ref. #641

mislav committed:
Import PR list API implementation to `pr/list` package

Also splits List vs. Search queries into separate methods for better
maintanability.

mislav committed:
Add `pr list --search`

mislav committed:
Add `pr list --author` filter

mislav committed:
Add tests for `listPullRequests`

GeorgeMac commented:
Thanks for looping me in @mislav 🙇

mislav committed:
WIP fix filter

mislav committed:
Unify checking whether search filters were passed by the user

samcoe commented on pkg/cmd/pr/list/http.go:
```diff
@@ -0,0 +1,247 @@
+	q.SetType(githubsearch.PullRequest)
+	q.InRepository(ghrepo.FullName(repo))
+	q.AddQuery(filters.Search)
+	q.SortBy(githubsearch.CreatedAt, githubsearch.Desc)
```
Will this override any sort parameter the user might have specified in `filters.Search`?

mislav replied:
That's a great question. Any user's query will always be appended at the end, so their own `sort:*` directive would override this implicit `sort:` directive after ElasticSearch parses it. However, after some thought, I decided that since `sort:created-desc` is already the default for issue search, it doesn't really need to be explicitly specified here. I'll remove it 👍

samcoe reviewed (approved):
👏 Nice work! I think searching for issues and prs is going to be a very popular feature.

samcoe commented on pkg/cmd/pr/shared/params.go:
```diff
@@ -153,6 +153,34 @@ type FilterOptions struct {
 	Search     string
 }
 
+func (opts *FilterOptions) IsDefault() bool {
```
This is much cleaner!

mislav committed:
Remove implicit `sort:created-desc` sort clause for `pr list`

This is the default sort mode for issues, so it's not needed to
explicitly set it. Furthermore, the user can specify their own sort mode
through the `--search` option.

Zayanusman221 commented:
> This adds search functionality to `pr list` to complement `issue list --search`.
> 
> Since the search backend is now easily available through code, this also adds support for `--author` filter for compatibility with `issue list` #2628 /cc @GeorgeMac
> 
> Followup to #3196 /cc @g14a
> 
> Fixes #452, ref. #641
